Implement simple wrapper over multiple TestProcessorNodes

camilbancioiu opened the pull request:
This PR adds the `TestNetwork` utility struct.

Can be used to replace this:
```
	advertiser := integrationTests.CreateMessengerWithKadDht("")
	_ = advertiser.Bootstrap(0)

	numOfShards := 1
	nodesPerShard := 1
	numMetaNodes := 10

	nodes := integrationTests.CreateNodes(
		numOfShards,
		nodesPerShard,
		numMetaNodes,
		integrationTests.GetConnectableAddress(advertiser),
	)
	integrationTests.DisplayAndStartNodes(nodes)

	defer func() {
		_ = advertiser.Close()
		for _, n := range nodes {
			_ = n.Messenger.Close()
		}
	}()
```
with this:
```
	net := integrationTests.NewTestNetwork(t)
	net.NumShards = 1
	net.NodesPerShard = 1
	net.NodesInMetashard = 10

	net.Start()
	defer func() {
		net.Close()
	}()
```

Nodes are available at `net.Nodes`.

Minting all the nodes: 
```
net.MintNodeAccounts(bigInt)
// or
net.MintNodeAccountsInt64(intValue)
```

Stepping over rounds:
```
net.Step()
net.Steps(multipleRounds)
```
`Step()` and `Steps()` do not return the round and nonce, because they are available at `net.Round` and `net.Nonce`, automatically updated at each step.

Please add more methods and make `TestNetwork` as useful as possible, to avoid copy-pasting initialization code for tests.

Any test-specific utility functions may be written to accept a `TestNetwork` instead of parameters such as `nodes`, `idxProposers`, `nonce`, `round` etc.

iulianpascalau commented on integrationTests/testNetwork.go:
```diff
@@ -0,0 +1,107 @@
+}
+
+func NewTestNetwork(t *testing.T) *testNetwork {
+	// TODO replace testing.T with testing.TB everywhere in the integrationTest utilities
```
:+1:

iulianpascalau commented on integrationTests/testNetwork.go:
```diff
@@ -0,0 +1,107 @@
+}
+
+func (net *testNetwork) Start() {
+	net.createAdvertiser()
```
:+1:

sasurobert commented on integrationTests/testNetwork.go:
```diff
@@ -0,0 +1,107 @@
+	net.Nonce = 0
+}
+
+func (net *TestNetwork) Step() {
```
add comment for exported public functions

iulianpascalau replied:
Right, missed that. :D

camilbancioiu replied:
Added.

camilbancioiu commented:
Added multiple improvements (see methods of `TestNetwork`).
Added comments.
Rewrote `TestTransaction_TransactionSCScenarios()` to use `TestNetwork` as demonstration.

SebastianMarian commented on integrationTests/testNetwork.go:
```diff
@@ -0,0 +1,407 @@
+}
+
+func (net *TestNetwork) handleOrBypassError(err error) {
+	if net.BypassErrorsOnce == true {
```
if net.BypassErrorsOnce {

camilbancioiu replied:
Remnant of a previous condition, removed.

iulianpascalau commented on integrationTests/testNetwork.go:
```diff
@@ -1,107 +1,407 @@
+	net.NodesSharded = make(NodesByShardMap)
+	for _, node := range net.Nodes {
+		shardID := node.ShardCoordinator.SelfId()
+		_, exists := net.NodesSharded[shardID]
```
L351-L354 can be safely removed

camilbancioiu replied:
Removed.

camilbancioiu commented on integrationTests/testNetwork.go:
```diff
@@ -1,107 +1,407 @@
 	"github.com/stretchr/testify/require"
 )
 
+type ShardIdentifier = uint32
```
Added.

camilbancioiu commented on integrationTests/testNetwork.go:
```diff
@@ -1,107 +1,407 @@
+
+// MintWallets adds the specified value to the test wallets.
+func (net *TestNetwork) MintWallets(value *big.Int) {
+	MintAllPlayers(net.Nodes, net.Wallets, value)
```
Added TODOs.

camilbancioiu commented on integrationTests/testNetwork.go:
```diff
@@ -1,107 +1,407 @@
+type NodesByShardMap = map[ShardIdentifier]NodeSlice
+type GasScheduleMap = map[string]map[string]uint64
+
+// TODO combine TestNetwork with the preexisting TestContext and OneNodeNetwork
```
Fixed.
